File: express/src/middleware/auth-middleware.js
import db from "../application/database.js";
import { ResponseError } from "../response/response-error.js";
import jwt from "jsonwebtoken";

export const authMiddleware = async (req, res, next) => {
  const token = req.get("Authorization");

  if (!token) {
    return next(new ResponseError(401, "Unauthorized"));
  }

  try {
    const decodedToken = jwt.verify(token, process.env.JWT_SECRET);
    req.user = decodedToken;

    const [rows] = await db
      .promise()
      .query("SELECT * FROM sessions WHERE token = ?", [token]);

    if (rows.length === 0) {
      return next(new ResponseError(401, "Unauthorized"));
    }

    next();
  } catch (error) {
    return next(new ResponseError(401, "Unauthorized"));
  }
};

File: flask/src/middleware/auth_middleware.py
import os
import jwt
from functools import wraps
from flask import request, jsonify
from application.database import get_connection
from response.response_error import ResponseError
from application.logging import logger

# Middleware untuk autentikasi
def auth_middleware(f):
    @wraps(f)
    def decorated_function(*args, **kwargs):
        token = request.headers.get("Authorization")
        logger.info(token)

        if not token:
            raise ResponseError(401, "Unauthorized")

        try:
            jwt_secret = os.getenv("JWT_SECRET")
            decoded_token = jwt.decode(token, jwt_secret, algorithms=["HS256"])

            connection = get_connection()
            if connection is None:
                raise ResponseError(500, "Error obtaining database connection")
            
            cursor = connection.cursor(dictionary=True)
            query = "SELECT * FROM sessions WHERE token = %s"
            cursor.execute(query, (token,))
            session = cursor.fetchone()

            if not session:
                raise ResponseError(401, "Unauthorized")
            
            request.user = decoded_token
            return f(*args, **kwargs)
        except jwt.ExpiredSignatureError:
            raise ResponseError(401, "Token has expired")
        except jwt.InvalidTokenError:
            raise ResponseError(401, "Invalid token")
        except ResponseError as e:
            raise e
        except Exception as e:
            raise ResponseError(500, str(e))
        finally:
            if cursor:
                cursor.close()
            if connection:
                connection.close()
    return decorated_function